ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/AuthHandlers.java

```diff
@@ -67,10 +67,16 @@ static void setupAuthHandlers(final Server server, final Router router,
               server.getConfig().getList(KsqlRestConfig.AUTHENTICATION_ROLES_CONFIG)),
           Provider.JAAS));
 
+      // pause before using the PLUGIN handlers because they can execute blocking code,
+      // which may cause the request to be read asynchronously
+      router.route().handler(AuthHandlers::pauseHandler);
+
       // set up using PLUGIN
       pluginHandler.ifPresent(h -> router.route().handler(wrappedHandler(h, Provider.PLUGIN)));
 
+      // pause again in case the plugin "resumed" the context on its own
```

Comment:
Just checking: is it safe to call pause on a context that is already paused, in the event that the pluginHandler doesn't resume the context on its own?

Reply:
Also, is it safe for a custom plugin handler implementation to resume the context? Even though we pause again here, I'm wondering if the fact that the context is unpaused for a short period of time in between would cause any async calls that have already been issued to start before we pause again. I wonder if it makes more sense to say that custom plugin handler implementations should not call resume on the context. (We could still leave this extra call here to be defensive, just trying to understand whether there are valid use cases for a plugin handler to call resume or not.)

Reply:
> Just checking: is it safe to call pause on a context that is already paused, in the event that the pluginHandler doesn't resume the context on its own?

yup, it's safe to pause multiple times

> Also, is it safe for a custom plugin handler implementation to resume the context? 

yup, so long as they don't do it asynchronously from the event loop context (that's the change in vertx 4.x that made me change this code in first place, the `BasicAuthHandler` (part of vertx library) does exactly that 😢 )

> I wonder if it makes more sense to say that custom plugin handler implementations should not call resume on the context. 

See above, that's unfortunately in the vertx default implementation

Reply:
The BasicAuthHandler is run as the `jaas` handler, before the first pause statement -- I don't think we ever run it as the custom `pluginHandler` between the first and second pause statements, right? Your point still stands though -- let's keep the defensive second pause statement.